Mathlib/AlgebraicTopology/ModelCategory/Cylinder.lean

```diff
@@ -0,0 +1,262 @@
+  cofibration_i : Cofibration P.i := by infer_instance
+
+/-- A good cylinder object `P` is very good if `P.π` is a (trivial) fibration. -/
+class IsVeryGood [HasBinaryCoproduct A A] [CategoryWithCofibrations C]
+    [CategoryWithFibrations C] : Prop extends P.IsGood where
+  fibration_π : Fibration P.π := by infer_instance
+
+attribute [instance] IsGood.cofibration_i IsVeryGood.fibration_π
+
+section
+
+variable [HasBinaryCoproduct A A] [CategoryWithCofibrations C]
+  [HasInitial C] [(cofibrations C).IsStableUnderComposition]
+  [(cofibrations C).IsStableUnderCobaseChange]
+  [IsCofibrant A] [P.IsGood]
+
+instance : Cofibration P.i₀ := by
+  rw [← P.inl_i]
+  infer_instance
+
+instance : Cofibration P.i₁ := by
+  rw [← P.inr_i]
+  infer_instance
+
+instance : IsCofibrant P.I :=
+  isCofibrant_of_cofibration P.i₀
+
+end
+
+instance [HasBinaryCoproducts C] [CategoryWithCofibrations C] [P.IsGood]
+    [(cofibrations C).RespectsIso] : P.symm.IsGood where
+  cofibration_i := by
+    have hi : cofibrations C P.i := by rw [← cofibration_iff]; infer_instance
+    rw [P.symm_i, cofibration_iff]
+    refine ((cofibrations C).arrow_mk_iso_iff ?_).2 hi
+    exact Arrow.isoMk (coprod.braiding A A) (Iso.refl _)
+
+section
+
+variable [CategoryWithCofibrations C] [CategoryWithFibrations C]
+  [(fibrations C).IsStableUnderComposition]
+
+instance [HasBinaryCoproduct A A] [HasTerminal C] [IsFibrant A] [P.IsVeryGood] : IsFibrant P.I :=
+  isFibrant_of_fibration P.π
+
+instance [(cofibrations C).RespectsIso] [HasBinaryCoproducts C] [P.IsVeryGood] :
+    P.symm.IsVeryGood where
+  fibration_π := by
+    dsimp
+    infer_instance
+
+end
+
+end
+
+variable [ModelCategory C] {A : C} (P : Cylinder A)
+
+section
+
+variable (h : MorphismProperty.MapFactorizationData (cofibrations C) (trivialFibrations C)
+    (codiag A))
+
+/-- A cylinder object for `A` can be obtained from a factorization of the obvious
+map `A ⨿ A ⟶ A` as a cofibration followed by a trivial fibration. -/
+@[simps]
+noncomputable def ofFactorizationData : Cylinder A where
+  I := h.Z
+  i₀ := coprod.inl ≫ h.i
+  i₁ := coprod.inr ≫ h.i
+  π := h.p
+
+@[simp]
+lemma ofFactorizationData_i : (ofFactorizationData h).i = h.i := by aesop_cat
+
+instance : (ofFactorizationData h).IsVeryGood where
+  cofibration_i := by simpa using inferInstanceAs (Cofibration h.i)
+  fibration_π := by dsimp; infer_instance
+
+instance [HasTerminal C] [IsFibrant A] [(fibrations C).IsStableUnderComposition] :
+    IsFibrant (ofFactorizationData h).I :=
+  isFibrant_of_fibration (ofFactorizationData h).π
+
+end
+
+variable (A) in
+lemma exists_very_good :
+    ∃ (P : Cylinder A), P.IsVeryGood :=
+  ⟨ofFactorizationData (MorphismProperty.factorizationData _ _ _),
+    inferInstance⟩
+
+instance : Nonempty (Cylinder A) := ⟨(exists_very_good A).choose⟩
+
+/-- The gluing of two good cylinders. -/
+@[simps]
+noncomputable def trans [IsCofibrant A] (P P' : Cylinder A) [P'.IsGood] :
```

Comment:
Since `symm` is defined for precylinders, perhaps we can also split the data part and the prop part here as well? defining `Precylinder.trans` from relevant `HasPushout` instance, and then having `Cylinder.trans` only prove that in good cases like this the precylinder is a cylinder?

Admittedly, for applications of precylinders where we don’t have yet a model category, transitivity is not very important (we’re working with a fixed cylinder), but since you’ve done symmetry for precylinders, might as well be consistent with this pattern.